File: datagen/generators/level23.py
"""Level 23 — The Market (final boss).

1000 days of GARCH(1,1) volatility clustering + Merton jumps + one hidden
regime shift (day 620: base variance x2.5, drift flips negative).

Distribution-mode grading: submit the 5th / 50th / 95th percentile of
tomorrow's close. Scored by mean pinball loss against 10,000 Monte Carlo
draws from the TRUE next-day distribution (model state known only to the
generator). lossMax = 1.10 x the loss of the exact true quantiles, so honest
uncertainty beats lucky point guesses. The intended pipeline — center at the
last price, width from EWMA/GARCH vol estimated on the current regime —
passes; an unconditional full-history vol fails (verified in tests).
"""

import numpy as np

from datagen.common import make_pack, rng, truth_block, _round_list

SEED = 3334  # chosen so the final state is hot: EWMA conditioning passes, flat history fails
N = 1000
DT = 1.0 / 252.0
A_G = 0.09          # GARCH arch coefficient
B_G = 0.88          # GARCH garch coefficient
OMEGA1 = 3.0e-6     # regime 1 base (uncond daily var 1e-4 -> vol 1.0%/day)
OMEGA2 = 7.5e-6     # regime 2: x2.5 (uncond vol ~1.58%/day)
MU1 = 0.0004
MU2 = -0.0004
SHIFT_DAY = 620
LAM = 6.0           # jumps / year
S_J = 0.05
S0 = 100.0
TAUS = [0.05, 0.5, 0.95]
N_DRAWS = 10_000


def simulate(g: np.random.Generator):
    """Returns (prices len N+1, final sigma2_next, last_price)."""
    r = np.empty(N)
    sig2 = OMEGA1 / (1 - A_G - B_G)
    eps_prev2 = sig2
    for i in range(N):
        omega = OMEGA1 if i < SHIFT_DAY else OMEGA2
        mu = MU1 if i < SHIFT_DAY else MU2
        sig2 = omega + A_G * eps_prev2 + B_G * sig2
        eps = np.sqrt(sig2) * g.normal()
        jump = 0.0
        k = g.poisson(LAM * DT)
        if k > 0:
            jump = g.normal(0.0, S_J * np.sqrt(k))
        r[i] = mu + eps + jump
        eps_prev2 = eps**2
    prices = S0 * np.exp(np.concatenate([[0.0], np.cumsum(r)]))
    sig2_next = OMEGA2 + A_G * eps_prev2 + B_G * sig2
    return prices, float(sig2_next)


def next_day_draws(last_price: float, sig2_next: float, g: np.random.Generator,
                   n: int = N_DRAWS) -> np.ndarray:
    z = g.normal(0.0, 1.0, n)
    k = g.poisson(LAM * DT, n)
    jump = np.where(k > 0, g.normal(0.0, 1.0, n) * S_J * np.sqrt(np.maximum(k, 1)), 0.0)
    r = MU2 + np.sqrt(sig2_next) * z + jump
    return last_price * np.exp(r)


def pinball(qs, taus, draws):
    loss = 0.0
    for q, tau in zip(qs, taus):
        d = draws - q
        loss += np.mean(np.where(d >= 0, tau * d, (tau - 1) * d))
    return loss / len(taus)


def build() -> dict:
    g = rng(SEED)
    prices, sig2_next = simulate(g)
    last = float(prices[-1])
    draws = next_day_draws(last, sig2_next, rng(SEED + 1))

    true_q = [float(np.quantile(draws, tau)) for tau in TAUS]
    loss_opt = pinball(true_q, TAUS, draws)
    loss_max = 1.10 * loss_opt

    t = np.arange(N + 1, dtype=float)
    return make_pack(
        id=23,
        seed=SEED,
        kind="series",
        columns={"t": t, "y": prices},
        truth=truth_block(
            {
                "omega1": OMEGA1, "omega2": OMEGA2, "alpha": A_G, "beta": B_G,
                "lambda": LAM, "sJ": S_J, "shift_day": float(SHIFT_DAY),
                "sigma_next_daily": float(np.sqrt(sig2_next)),
            },
            r"r_t = \mu_{\text{reg}} + \sigma_t z_t + J\,dN_t,\quad \sigma_t^2 = \omega_{\text{reg}} + 0.09\,\varepsilon_{t-1}^2 + 0.88\,\sigma_{t-1}^2,\quad \text{regime shift at } t = 620",
            states=[0] * (SHIFT_DAY + 1) + [1] * (N - SHIFT_DAY),
        ),
        grading={
            "mode": "distribution",
            "taus": TAUS,
            "draws": _round_list(draws),
            "lossMax": float(f"{loss_max:.8g}"),
            "trueQuantiles": [float(f"{q:.8g}") for q in true_q],
        },
    )

File: src/levels/meta/level23.ts
import type { LevelMeta } from '../../lib/types';

const meta: LevelMeta = {
  id: 23,
  slug: 'the-market',
  title: 'The Market',
  tier: 'QUANT GAUNTLET',
  tagline: 'final boss. predict tomorrow — as a distribution, like an adult.',
  brief: String.raw`One thousand days of a full synthetic market: **GARCH(1,1)** volatility clustering, **Merton jumps**, and **one hidden regime shift** somewhere in the record. Everything you've cleared was training for this chart.

The task is not to predict tomorrow's close. Nobody can, and the grader knows it. The task is to state tomorrow's close **as a distribution**: submit the 5th, 50th, and 95th percentiles of $S_{T+1}$.

Scoring is mean pinball loss against 10,000 hidden Monte-Carlo draws from the *true* next-day distribution — the one conditioned on the actual model state. The bar is 110% of the loss achieved by the exact true quantiles. A lucky point guess cannot pass; an honest, well-conditioned uncertainty band passes with room. Bring: case 5 (differencing), case 7 (log-returns), case 15 (jumps), case 17 (regimes).`,
  recover: 'Submit the 5th / 50th / 95th percentiles of tomorrow’s close.',
  submitNote: 'Quantiles must be non-decreasing.',
  starterCode: String.raw`# GROUND TRUTH · CASE 23 — The Market
s = data.y.to_numpy()
r = np.diff(np.log(s))
last = s[-1]
print("last close:", round(last, 3))

# exhibit 1: vol clusters. exhibit 2: something changed mid-record.
roll = pd.Series(r).rolling(21).std().to_numpy() * np.sqrt(252)
overlay(data.t.to_numpy()[1:], roll * last / np.nanmax(roll) , label="rolling vol (scaled)", style="dot")

# the grader scores three quantiles of TOMORROW's close.
# history matters only insofar as it describes tomorrow.
`,
  hints: {
    nudge: String.raw`Center near the last price (case 7 said drift is unknowable at this horizon). All the skill is in the **width** — and the width must reflect *current* volatility, not the 1000-day average. The record has two eras; only the recent one is tomorrow's.`,
    method: String.raw`EWMA (RiskMetrics) vol: $v_t = \lambda v_{t-1} + (1-\lambda) r_t^2$ with $\lambda \approx 0.94$, giving $\hat\sigma_{T+1}$ that automatically forgets the old regime. Then

$$\hat q_\tau = S_T\, e^{\hat\sigma_{T+1}\, \Phi^{-1}(\tau)}$$

for $\tau = 0.05, 0.5, 0.95$. Refinements that also fit within the bar: fit actual GARCH(1,1) by MLE on the recent era for a proper one-step forecast; widen the 5% tail a touch for jump risk (case 15). Overkill that won't help: any attempt to forecast the *sign* of tomorrow.`,
    derivation: String.raw`**Why pinball loss makes honesty optimal**: the expected pinball loss $\mathbb{E}[\ell_\tau(q)]$ is uniquely minimized at the true $\tau$-quantile — it is a *proper scoring rule* for quantiles. Differentiating $\mathbb{E}[(\tau - \mathbb{1}\{X < q\})(X - q)]$ gives the first-order condition $P(X < q) = \tau$. There is no strategy that beats reporting your true beliefs; the only way to score well is to have good beliefs. That is the entire ideology of this game in one loss function.

**Why EWMA is near-optimal here**: GARCH(1,1) one-step variance is $\sigma_{T+1}^2 = \omega + \alpha \varepsilon_T^2 + \beta \sigma_T^2$; with persistence $\alpha + \beta \approx 0.97$, the model is close to IGARCH, whose exact solution *is* an EWMA of squared returns. The regime shift is handled for free: EWMA's effective memory $\approx 1/(1-\lambda) \approx 17$ days barely sees the old era. The test suite verifies by Monte Carlo that this pipeline scores within the 110% band of the true quantiles — and that a full-history flat-σ band **fails**, because this particular market ends hot: the true next-day σ is roughly twice the unconditional estimate.

**Why the median is easy and the tails are the game**: the median contributes $\approx \mathbb{E}|X - m|/3$ of the loss regardless of your vol estimate; the 5/95 quantiles are where a wrong σ bleeds score quadratically. Under-dispersed intervals (the classic sin: standing in front of a regime shift with a long-run σ) get punished on every tail draw that escapes.

You have completed the syllabus: means (1–2), spectra (3–4, 8–12), stochastic processes (5–7, 13–19), and the multiple-hypothesis honesty of 20–22, all compiled into three numbers and a calibrated shrug. That shrug — *the correct width* — is the most employable object in quantitative finance. GROUND TRUTH · CASE FILE CLOSED.`,
  },
  xLabel: 't (days)',
  yLabel: 'close',
};

export default meta;
